README.md
emo - E-Mail Obfuscation
================================================================================

E-Mail Obfuscation with Javascript for the MODX Evolution content management
framework

This plugin replaces all plaintext emails and email links with span elements.
Then the email addresses are encrypted and stored in javascript variables.
Javascript decryption routines are triggered by the browser on 'window.onload'
event and replace the span elements with the real email addresses.

Features:
--------------------------------------------------------------------------------
This plugin searches for 'mailto:' strings in href attributes and all plain text
email addresses in the html output. However, only email addresses like
'user@host.com' are matched whereas 'Tog@ther' is left untouched.

Matches of all plaintext emails and email links are consequently replaced with
span elements only containing a note to turn on javascript. Then email addresses
and original link text are encrypted and stored in javascript variables.
Decryption routines triggered by the browser on 'window.onload' event are
located in the script file 'emo.js'. Optionally, CSS class 'emo_address' can be
used to configure the appearance of email links.

It is still believed that hexadecimal or unicode encoding will stop spam-bots
being able to find your email address. Nevertheless, encoded email address
harvester are on the way. Unlike other obfuscation plugins, this one uses real
encryption instead of using an out-dated and over-used hack such as hexadecimal
or unicode encoding (not encryption) of email addresses. Also, all traces of
href attributes as well as 'mailto:' strings and '@' characters are hidden
from spam-bots. But it does not modify adresses inside &lt;form&gt; tags. So posted
and not validated forms do not break.

If the 'No javascript' message contains a number, a link to a MODX resource with
that ID is generated. This could i.e. point to a resouce with a contact form.

Installation:
--------------------------------------------------------------------------------
1. Upload the folder *assets/plugins/emo* in the corresponding folder in your installation.
2. Create a plugin called 'emo' with the plugin code `include MODX_BASE_PATH.'/assets/plugins/emo/emo.plugin.php';` and check the `OnWebPagePrerender` and `OnLoadWebDocument` events.
3. Insert the following code into the *Plugin configuration*

```
&emoTplOnly=Templates only?;list;enable,disable;enable
&emoSelection=How to select IDs?;list;exclude,include;exclude
&emoSelectionRange=Chosen IDs;text;
&pathToEmoJs=Path to emo javascript;text;/assets/plugins/emo/emo.min.js
&pathToEmoCss=Path to emo css;text;
&noScriptMessage='No javascript' message (Resource ID for internal link);text;Turn on JavaScript!
```

Plugin Settings:
--------------------------------------------------------------------------------

Property | Description | Default
---- | ----------- | -------
emoTplOnly | Don't work on resources with blank template | Yes
emoSelection | Selection type for enabled/disabled resources for emo | -
emoSelectionRange | Comma separated list of enabled/disabled resource IDs for emo |
pathToEmoJs | Path to emo javascript | /assets/components/emo/js/emo.min.js
pathToEmoCss | Path to emo css |
noScriptMessage | 'No javascript' message (Resource ID for internal link) | Turn on JavaScript!


HISTORY.md
History
================================================================================

- 1.5.1 Novemeber 26, 2013
    - Bugfix for regClient of emo.js

- 1.5 October 25, 2013
    - Package code is now folder independent

- 1.4 January 5, 2011
    - External emo class file
    - Does not modify adresses inside <form> tags, works better with mailto hrefs
    - Call javascript file by regClientScript
    - Noscript message could be set to an ID - a link to this page will be generated

- 1.3 January 19, 2008
    - Custom Base64 encoding

- 1.2 January 17, 2008
    - Email encoding

- 1.1 January 12, 2008
    - First public release


assets/plugins/emo/emo.js
/*
 *  Written by: Florian Wobbe
 *  Contact: www.eprofs.de
 *  Created: 2008-01-12 - v1.1
 *  Modified: 2008-01-17 - v1.2 - Email encoding
 *  Modified: 2008-01-19 - v1.3 - Custom Base64 encoding
 *  For: MODx cms (modxcms.com)
 *  Name: emo - E-Mail Obfuscation
 *  Description: These javascript functions replace html placeholder
 *    span elements with decrypted email addresses.
 *  License: GPL
 */

// addLoadEvent is Simon Willison's function that allows for convenient
// addition of multiple fuctions that are all supposed to be triggered
// on window.onload event.
// takes function name as the argument. e.g. addLoadEvent(doPopups);
function addLoadEvent(func) {
	var oldonload = window.onload;
	if (typeof window.onload !== 'function') {
		window.onload = func;
	} else {
		window.onload = function() {
			if (oldonload) {
				oldonload();
			}
			func();
		};
	}
}

// Decrypt all email addresses
function emo_replace() {
	for (var i = 1; i < emo_addresses.length; i++) {
		var id = '_emoaddrId' + i;
		var elem = document.getElementById(id);
		if (elem) {
			if (elem.firstChild) {
				elem.removeChild(elem.firstChild);
			}
			elem.innerHTML = decrypt_string(i);
		}
	}
}

// Manage decryption cache
var decryption_cache = new Array();
function decrypt_string(n) {
	var cache_index = "'" + n + "'";
	if (decryption_cache[cache_index]) {
		// If this string has already been decrypted, just return the cached version.
		return decryption_cache[cache_index];
	}
	if (emo_addresses[n]) {
		// Is crypted_string an index into the addresses array?
		var crypted_string = emo_addresses[n];
	}
	if (!crypted_string.length) {
		// Make sure the string is actually a string
		return "Error, not a valid index.";
	}
	var decrypted_string = decode_base64(crypted_string);
	// Cache this string for any future calls
	decryption_cache[cache_index] = decrypted_string;
	return decrypted_string;
}

// Custom base 64 decoding
function decode_base64(data) {
	var tab = emo_addresses[0];
	var out = "", c1, c2, c3, e1, e2, e3, e4;
	for (var i = 0; i < data.length; ) {
		e1 = tab.indexOf(data.charAt(i++));
		e2 = tab.indexOf(data.charAt(i++));
		e3 = tab.indexOf(data.charAt(i++));
		e4 = tab.indexOf(data.charAt(i++));
		c1 = (e1 << 2) + (e2 >> 4);
		c2 = ((e2 & 15) << 4) + (e3 >> 2);
		c3 = ((e3 & 3) << 6) + e4;
		out += String.fromCharCode(c1);
		if (e3 !== 64)
			out += String.fromCharCode(c2);
		if (e4 !== 64)
			out += String.fromCharCode(c3);
	}
	return out;
}


assets/plugins/emo/emo.class.php
<?php

/**
 * emo
 *
 * Copyright 2008-2011 by Florian Wobbe - www.eprofs.de
 * Copyright 2011-2013 by Thomas Jakobi <thomas.jakobi@partout.info>
 *
 * emo is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option) any
 * later version.
 *
 * emo is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * emo; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * @package emo
 * @subpackage classfile
 */
class Emo {

	/**
	 * A reference to the modX instance
	 * @var modX $modx
	 */
	public $modx;

	/**
	 * Address counter
	 * @var int $addrCount
	 */
	public $addrCount;

	/**
	 * A string of debug informations
	 * @var string $debug
	 */
	public $debug;

	/**
	 * A string for storing the javascript
	 * @var string $addressesjs
	 */
	public $addressesjs;

	/**
	 * A configuration array
	 * @var array $config
	 */
	private $config;

	/**
	 * The no script message
	 * @var string $noScriptMessage
	 */
	private $noScriptMessage;

	/**
	 * An Array for storing recent links
	 * @var array $recentLinks
	 */
	private $recentLinks;

	/**
	 * base 64 characters
	 * @var string $tab
	 */
	private $tab;

	/**
	 * emo constructor
	 *
	 * @access public
	 * @param modX $modx A reference to the modX instance.
	 * @param array $params An array of configuration options.
	 */
	public function __construct(&$modx, $params) {
		$this->modx = & $modx;
		$this->noScriptMessage = $params['noScriptMessage'];
		$this->tab = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+.';
		$this->recentLinks = array();
		$this->config = array(
			'show_debug' => (bool) isset($params['show_debug']) ? $params['show_debug'] : FALSE
		);
	}

	/**
	 * Create regular expression for searching email addresses.
	 * Modified from ObfuscateEmail plugin for MODX Evolution by Aloysius Lim.
	 *
	 * @access private
	 * @param void
	 * @return string Regular expression.
	 */
	private function emailRegex() {
		$atom = "[-!#$%'*+/=?^_`{|}~0-9A-Za-z]+";
		$email_left = $atom . '(?:\\.' . $atom . ')*';
		$email_right = $atom . '(?:\\.' . $atom . ')+';
		$email = $email_left . '@' . $email_right;
		return $email;
	}

	/**
	 * Custom base 64 encoding
	 * Original emo code by Florian Wobbe - www.eprofs.de
	 *
	 * @access public
	 * @param string $data String to encode.
	 * @return string Encoded data
	 */
	private function encodeBase64($data) {
		$out = '';
		for ($i = 0; $i < strlen($data);) {
			$c1 = ord($data {$i++});
			$c2 = $c3 = NULL;
			if ($i < strlen($data))
				$c2 = ord($data {$i++});
			if ($i < strlen($data))
				$c3 = ord($data {$i++});
			$e1 = $c1 >> 2;
			$e2 = (($c1 & 3) << 4) + ($c2 >> 4);
			$e3 = (($c2 & 15) << 2) + ($c3 >> 6);
			$e4 = $c3 & 63;
			if (is_nan($c2))
				$e3 = $e4 = 64;
			else if (is_nan($c3))
				$e4 = 64;
			$out .= $this->tab {$e1} . $this->tab {$e2} . $this->tab {$e3} . $this->tab {$e4};
		}
		return $out;
	}

	/**
	 * Encrypt the match or generate a link when linktext is missing
	 * Modified original emo code by Florian Wobbe - www.eprofs.de
	 *
	 * @access public
	 * @param string $matches String to encode.
	 * @return string Encoded data
	 */
	private function encodeLink($matches) {
		// Use global variables

		if (!$this->addrCount) {
			// Random generator seed
			mt_srand((double) microtime() * 1000000);
			// Make base 64 key
			$this->tab = str_shuffle($this->tab);
			// Set counter and add base 64 key to array
			$this->addrCount = 0;
			$this->addressesjs .= '      emo_addresses[' . $this->addrCount++ . '] = "' . $this->tab . '";' . "\n";
		}

		// Link without a linktext: insert email address as text part
		if (sizeof($matches) < 3) {
			$matches[2] = $matches[1];
		}

		// urlencode a possible subject
		$matches[1] = preg_replace('!(.*\?(subject|body)=)([^\?]*)!ie', "'$1'.rawurldecode(rawurlencode('$3'))", $matches[1]);

		// Create html of the true link
		$trueLink = '<a class="emo_address" href="mailto:' . $matches[1] . '">' . $matches[2] . '</a>';

		// Did we use the same link before?
		$key = array_search($trueLink, $this->recentLinks);
		if ($key === false) {
			// Encrypt the complete link
			$crypted = '"' . $this->encodeBase64($trueLink) . '"';
		} else {
			// Use previously encrypted link
			$crypted = 'emo_addresses[' . ($key + 1) . ']';
		}

		// Add encrypted address to array
		$this->addressesjs .= '      emo_addresses[' . $this->addrCount . '] = ' . $crypted . ';' . "\n";

		// Create html of the fake link
		$replaceLink = '<span id="_emoaddrId' . $this->addrCount . '"><span class="emo_address">' . $this->noScriptMessage . '</span></span>';

		// Add link to recent links array
		array_push($this->recentLinks, $trueLink);

		// Debugging
		if ($this->config['show_debug']) {
			$this->debug .= '  ' . $this->addrCount . ' ' . $matches[0] . "\n" .
					'    ' . $matches[1] . "\n" .
					'    ' . $matches[2] . "\n" .
					'    ' . $crypted . "\n";
		}

		// Increase address counter
		$this->addrCount++;

		return $replaceLink;
	}

	/**
	 * Replace the found email strings and generate the javascript
	 * Modified original emo code by Florian Wobbe - www.eprofs.de
	 *
	 * @access public
	 * @param string $content String to encode.
	 * @return string Encoded data
	 */
	public function obfuscateEmail($content) {

		// Script block header
		$this->addressesjs = "\n" . '    <!-- This script block stores the encrypted //-->' . "\n" .
				'    <!-- email address(es) in an addresses array. //-->' . "\n" .
				'    <script type="text/javascript">' . "\n" . '    /* <![CDATA[ */' . "\n" .
				'      var emo_addresses = new Array();' . "\n";

		// Debugging
		if ($this->config['show_debug']) {
			$this->debug = "\n" . '<!-- Emo debugging' . "\n";
			$mtime = microtime();
			$mtime = explode(' ', $mtime);
			$mtime = $mtime[1] + $mtime[0];
			$starttime = $mtime;
		}

		// exclude form tags
		$splitEx = "#((?:<form).*(?:</form>))#isUu";
		$parts = preg_split($splitEx, $content, NULL, PREG_SPLIT_DELIM_CAPTURE);
		$output = '';
		foreach ($parts as $part) {
			if (substr($part, 0, 5) != '<form') {
				$part = preg_replace_callback('#<a[^>]*mailto:([^\'"]+)[\'"][^>]*>(.*)</a>#iUu', array($this, 'encodeLink'), $part);
				$part = preg_replace_callback('<(' . $this->emailRegex() . ')>', array($this, 'encodeLink'), $part);
			}
			$output .= $part;
		}

		// Finish encrypted addresses block
		$this->addressesjs .= '      addLoadEvent(emo_replace());' . "\n" .
				'    //-->' . "\n" .
				'    </script>' . "\n";

		// Maybe you want to use jQuery ...
		// $this->addressesjs .= '     $(window).load(function(){'."\n".'        emo_replace();'."\n".'      });'."\n".'    /* ]]> */'."\n".'    </script>'."\n";
		// Debugging
		if ($this->config['show_debug']) {
			$mtime = microtime();
			$mtime = explode(' ', $mtime);
			$mtime = $mtime[1] + $mtime[0];
			$endtime = $mtime;
			$totaltime = ($endtime - $starttime);
			$this->debug .= '  Email crypting took ' . $totaltime . ' seconds' . "\n\n" .
					'  ' . implode("\n  ", $this->recentLinks) . "\n" .
					'-->';
		}
		return $output;
	}

}

?>


assets/plugins/emo/emo.plugin.php
<?php
/**
 * emo
 *
 * Copyright 2008-2011 by Florian Wobbe - www.eprofs.de
 * Copyright 2011-2013 by Thomas Jakobi <thomas.jakobi@partout.info>
 *
 * emo is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option) any
 * later version.
 *
 * emo is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * emo; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * @package emo
 * @subpackage pluginfile
 */
if (MODX_BASE_PATH == '') {
	die('<h1>ERROR:</h1><p>Please use do not access this file directly.</p>');
}

// set customtv (base) path
define('EMO_PATH', str_replace(MODX_BASE_PATH, '', str_replace('\\', '/', realpath(dirname(__FILE__)))) . '/');
define('EMO_BASE_PATH', MODX_BASE_PATH . EMO_PATH);

if (!class_exists('Emo')) {
	include EMO_BASE_PATH . 'emo.class.php';
}

// Get plugin settings
$emoTplOnly = isset($emoTplOnly) ? $emoTplOnly : 'disable';
$emoSelection = isset($emoSelection) ? $emoSelection : 'exclude';
$emoSelectionRange = isset($emoSelectionRange) ? $emoSelectionRange : '';
$pathToEmoJs = isset($pathToEmoJs) ? $pathToEmoJs : '/assets/plugins/emo/emo.js';
$pathToEmoCSS = isset($pathToEmoCSS) ? $pathToEmoCSS : '';
$noScriptMessage = isset($noScriptMessage) ? $noScriptMessage : 'Turn on Javascript!';
if (is_numeric($noScriptMessage)) {
	$doc = $modx->getDocument($noScriptMessage);
	$noScriptMessage = '<a href="' . $modx->makeUrl($doc['id']) . '">' . $doc['pagetitle'] . '</a>';
}

// Stop plugin on selection range and selection type
$emoSelectionRange = explode(',', str_replace(' ', '', $emoSelectionRange));
$emoFound = in_array($modx->documentObject['id'], $emoSelectionRange);
if (($emoFound && ($emoSelection != 'include')) || ($emoTplOnly == 'enable' && ($modx->documentObject['template'] == 0))) {
	return;
}

$e = &$modx->Event;
switch ($e->name) {
	case 'OnLoadWebDocument': {
			if ($pathToEmoJs != '') {
				$modx->regClientScript($pathToEmoJs);
			}
			if ($pathToEmoCSS != '') {
				$modx->regClientCSS($pathToEmoCss);
			}
			break;
		}
	case 'OnWebPagePrerender': {
			$params = array(
				'noScriptMessage' => $noScriptMessage,
				'show_debug' => $show_debug
			);
			$emo = new Emo($modx, $params);
			$modx->documentOutput = $emo->obfuscateEmail($modx->documentOutput);
			$script = $emo->addressesjs . $emo->debug;
			if ($emo->addrCount && strpos($modx->documentOutput, $script) === false) {
				$modx->regClientScript($script);
				$modx->documentOutput = preg_replace('~(</body[^>]*>)~i', $script . "\n" . '\1', $modx->documentOutput);
			}
			break;
		}
	default: {
			break;
		}
}
return;
?>
